misc/pragmas.md

```diff
@@ -0,0 +1,550 @@
+orthogonal.
+
+* What information should be encodable with pragmas?
+
+* What syntax should pragmas have in invisible XML? 
+
+* What representation should pragmas have in the XML serialization of
+an ixml grammar?
+
+* Where can pragmas appear?
+
+
+## Proposal(s)
+
+The current proposal is given the arbitrary name of 'brackets QName' for
+discussion; an earlier proposal (the 'hash-QName' proposal) has been
+withdrawn, though traces of it may remain in other documents in this
+branch.
+
+### The brackets-QName proposal 
+
+In this proposal, pragmas take the form of a left square bracket, an
+optional mark, a QName, the pragma's data, and a right square bracket.
+Nested pairs of square brackets are allowed, so pragmas can nest
+arbitrarily deep.
+
+In the ixml form of a grammar, pragmas can occur immediately
+
+* before a terminal or nonterminal symbol, before the mark if any, or
+
+* in the whitespace immediately before the full stop of a rule.
+
+In the XML form, the serialization of the pragma depends on its
+position and in some cases on the mark following the opening bracket.
+
+* Pragmas occurring before the full stop of a rule are serialized as
+extension elements following the final `alt` of the right-hand side of
+the rule, regardless of the mark.
+
+* Pragmas occurring before a symbol are
+
+    * serialized as attributes if marked `@` (ignoring marks on any
+    nested pragmas).
+
+    * serialized as processing instructions if marked `-`.
+
+      The processing instructions appear as the first children of the
+      XML representation of the symbol; the name and data of the
+      processing instruction in the XML form are the name and pragma
+      data of the pragma in the ixml form.
+
+    * serialized as extension elements if unmarked or marked `^`.
+
+      The extension elements appear as the first children of the XML
+      representation of the symbol. In the canonical XML
+      representation, the pragma data is serialized as a 'pragma-data'
+      element within the extension element, which will contain the
+      pragma-data as a character sequence. Depending on the semantics
+      of the pragma itself, the extension element may also contain
+      additional elements. (See the examples in the final section of
+      this paper.)
+
+For example:
+
+````
+    [my:color blue] a = b, [@my:flavor vanilla] c? [my:spin ...].
+````
+
+The corresponding XML form is:
+
+````
+    <rule name="a">
+        <my:color>blue</my:color>
+        <alt>
+            <nonterminal name="b"/>
+            <option>
+                <nonterminal name="c" my:flavor="vanilla"/>
+            </option>
+        </alt>
+        <my:spin>
+            <pragma-data>...</pragma-data>
+        </my:spin>
+    </rule>
+````
+
+Annotations appearing immediately before an occurrence of a symbol in
+the right-hand side of a rule pertain to the occurrence of that
+symbol.  In the example, this is the case for the `my:flavor` pragma
+on *c*.
```

Comment:
Should we say whether a pragma for a given non-terminal should be preferred for an instance of the same pragma on the corresponding rule?  (copied from a previous review)

Reply:
We agreed that this is for the pragma developer rather than the specification